lib/Dialect/TTMetal/Transforms/Passes.cpp

```diff
@@ -16,44 +16,324 @@
+    auto dm = calculateDataMovement(
+        inputTy.getShape(), inputLayout.getElementSizeBytes(), src, dst);
+
+    auto noc0Attr =
+        rewriter.getAttr<ttkernel::ThreadTypeAttr>(ttkernel::ThreadType::Noc0);
+    SmallVector<Attribute> threadTypes(dm.size(), noc0Attr);
+    SmallVector<Attribute> operand_cb_port_mapping;
+    SmallVector<Attribute> coreRanges;
+    coreRanges.reserve(dm.size());
+    for (auto [dstCoord, srcs] : dm) {
+      SmallVector<int64_t> offset = {dstCoord.y, dstCoord.x};
+      SmallVector<int64_t> size = {1, 1};
+      coreRanges.push_back(
+          rewriter.getAttr<ttmetal::CoreRangeAttr>(offset, size));
+    };
+
+    auto metalDispatch = rewriter.create<ttmetal::DispatchOp>(
+        op.getLoc(), SmallVector<Type>({outputTy}),
+        SmallVector<Value>({op.getInput()}),
+        SmallVector<Value>({op.getOutput()}), rewriter.getArrayAttr(coreRanges),
+        rewriter.getArrayAttr(threadTypes),
+        rewriter.getArrayAttr(operand_cb_port_mapping), threadTypes.size());
+
+    int i = 0;
+    PhysicalCoreCoordMapping physicalCoordMapping(systemDesc.getChipDescs());
+    std::int64_t inputBaseAddress = lookupAddress(op.getInput());
+    std::int64_t outputBaseAddress = lookupAddress(op.getOutput());
+    assert(inputBaseAddress);
+    assert(outputBaseAddress);
+    assert(inputBaseAddress % addressAlignment == 0);
+    assert(outputBaseAddress % addressAlignment == 0);
+    for (auto [dstCoord, srcs] : dm) {
+      Block *nocBlock = rewriter.createBlock(&metalDispatch.getRegion(i++));
+      OpBuilder nocBuilder(nocBlock, nocBlock->begin());
+      for (auto s : srcs) {
+        assert(s.srcOffset % addressAlignment == 0);
+        assert(s.dstOffset % addressAlignment == 0);
+        assert(s.size % addressAlignment == 0);
+        auto [yPhys, xPhys] = physicalCoordMapping[s.srcCoord];
+        auto y = nocBuilder.create<arith::ConstantOp>(
+            op.getLoc(), nocBuilder.getI32Type(),
+            nocBuilder.getI32IntegerAttr(yPhys));
+        auto x = nocBuilder.create<arith::ConstantOp>(
+            op.getLoc(), nocBuilder.getI32Type(),
+            nocBuilder.getI32IntegerAttr(xPhys));
+        auto srcOffset = nocBuilder.create<arith::ConstantOp>(
+            op.getLoc(), nocBuilder.getI32Type(),
+            nocBuilder.getI32IntegerAttr(inputBaseAddress + s.srcOffset));
+        auto srcRemoteNocAddr = nocBuilder.create<ttkernel::GetNocAddrOp>(
+            op.getLoc(), x, y, srcOffset);
+        auto dstLocalL1Addr = nocBuilder.create<arith::ConstantOp>(
+            op.getLoc(), nocBuilder.getI32Type(),
+            nocBuilder.getI32IntegerAttr(outputBaseAddress + s.dstOffset));
+        auto size = nocBuilder.create<arith::ConstantOp>(
+            op.getLoc(), nocBuilder.getI32Type(),
+            nocBuilder.getI32IntegerAttr(s.size));
+        nocBuilder.create<ttkernel::NocAsyncReadOp>(
+            op.getLoc(), srcRemoteNocAddr, dstLocalL1Addr, size);
```

Comment:
Let's move this block of code to a separate function `buildNocAsyncReadOp` just for the sake of readability.